### ISSUE TITLE
Support modeling health checks into the AppHost model for use in dashboard, deployment manifest, etc.

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/aspire#890]
We should consider adding support for modeling health checks endpoints into the AppHost app model so that resources' health checks endpoints can be integrated into the dashboard (#889) and emitted into the deployment manifest for potential auto-configuration by deployment tools, e.g. `azd` could configure [HTTP health probes when deploying Aspire apps to ACA](https://learn.microsoft.com/en-us/azure/container-apps/health-probes?tabs=arm-template#http-probes).

**Comment by mitchdenny** [comment]
Some API ideas:

```csharp
var builder = DistributedApplication.CreateBuilder(args);
var app = builder.AddProject<Projects.Foo>(...);
app.WithProbe(app.GetEndpoint("http"), "/health");
```

Signature would look like:

```csharp
public static IResourceBuilder<T> WithProbe<T>(this IResourceBuilder<T> builder, EndpointReference endpoint, string path) where T: IResourceWithHealthProbes
```

This would ultimately result in a `HealthProbeAnnotation` being added.

**Comment by davidfowl** [comment]
We need the manifest model as well. We also need to update the shared defaults to read the endpoint from config (today we hard code it)

**Comment by mitchdenny** [comment]
@davidfowl @DamianEdwards confirming this is still in for GA ... if so, this would be a good one to jump on @JamesNK.

**Comment by davidfowl** [comment]
No, this is post GA

**Comment by GMouaad** [comment]
I started some draft work on this issue on #4151 if someone would like to take a look at and provide some feedback.

Open points: 
- Probe endpoint types:
The draft supports TCP and http Probes. Is supporting commands necessary? it would make stuff more complicated.
- Additional parameters (InitialDelay and Period) and their defaults
I added some additional parameters to be used by some target platforms (K8s, ACA, etc.). Any thoughts on these and their defaults (5 Seconds on both)
- Extension with a callback
Can be quickly added if needed (?).
- Which resources should already implement the `IResourceWithHealthProbes`
In the tests I only use a dummy resource for now to showcase the functionality. 
I can implement it on an actual resource as a starting point maybe. Any suggestions on which one?

CC @mitchdenny @davidfowl Feel free to give feedback ;)

**Comment by mitchdenny** [comment]
@davidfowl @DamianEdwards @vhvb1989 @prom3theu5

Interested in your thoughts on the manifest structure here? We could introduce another field at the resource level for probes. But health probes are heavily correlated with endpoints/bindings so perhaps it makes more sense to make it a sub-property which might make for easier correlation by deployment tools?

To that end, I am wondering whether the probe API surface in the app model should be tied to endpoints as well (so more than just a manifest serialization issue).

**Comment by DamianEdwards** [comment]
While I understand the motivation, tying it to endpoints so closely feels a bit too limiting/messy. In the manifest, it's easy enough to use an expression to refer to the endpoint the probe utilizes, but I think there will likely be a few other properties required to properly model a probe such that keeping them separate is warranted, e.g. probe kind (health, liveness, readiness), mechanism (HTTP, TCP, executable, etc.), delay, HTTP-specific properties like path, status code, headers, body text, etc.

**Comment by wertzui** [comment]
Has there been any progress on this? I would really like to reflect my health probes for my projects in the Dashboard.

**Comment by thompson-tomo** [comment]
I am of the thought that we should try and standardise health probes to leverage the OTEL data model to capture the result of the health checks. 

I have raised https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1855 to discuss the concept of capturing the probe results as an event in OTEL. This approach would mean that aspire & other tools would simply provide a visualisation of those events. I have also raised for dotnet to provide an implementation which completes a periodic check & log the result.

**Comment by mitchdenny** [comment]
@samsp-msft

**Comment by julioct** [comment]
Any news on this one? Assuming I have enabled the health endpoints for Prod in Service Defaults, how do I tell my AppHost to turn on the liveness and readiness probes in ACA for my projects?

![image](https://github.com/user-attachments/assets/70b6bfff-5b2f-459c-be0f-5af5c9585b5a)

**Comment by davidfowl** [comment]
We’re working on health checks at the moment but there’s no active plan to work on health probes

**Comment by DamianEdwards** [comment]
> Any news on this one? Assuming I have enabled the health endpoints for Prod in Service Defaults, how do I tell my AppHost to turn on the liveness and readiness probes in ACA for my projects?

Today you'd have to do that manually. As @davidfowl said we haven't got active plans for modeling health probes into the app as yet, but as we make progress on the story for customizing ACA configuration from the App Host C# code, something might emerge.

**Comment by julioct** [comment]
Yes, I ended up having to `azd infra synth` and add this block to each of my 7 microservices:

```yaml
  template:
    containers:
      - image: {{ .Image }}
        name: catalog-service
        probes:
          - type: liveness
            httpGet:
              path: "/health/alive"
              port: 8081
              scheme: HTTP
            initialDelaySeconds: 10
            periodSeconds: 10
          - type: readiness
            httpGet:
              path: "/health/ready"
              port: 8081
              scheme: HTTP
            initialDelaySeconds: 10
            periodSeconds: 5
```

Sad, but works.

**Comment by Dona278** [comment]
As suggested by @davidfowl I move here the proposal from [here](https://github.com/dotnet/aspire/discussions/11068#discussioncomment-14186981)

Since we can customize publish output (e.g. kubernetes) may be useful to identify the kind of the health check and provide additional data.

Example of annotations (only some properties)
```cs
internal enum EndpointHealthCheckKind
{
    Liveness,
    Readiness,
}

internal abstract class EndpointHealthCheckAnnotation(string key, string endpointName, EndpointHealthCheckKind kind) : HealthCheckAnnotation(key)
{
    /// <summary>
    /// Type of the health check.
    /// </summary>
    public EndpointHealthCheckKind Kind { get; } = kind;

    /// <summary>
    /// Name of the endpoint to use for health check.
    /// </summary>
    public string EndpointName { get; } = endpointName;

    /// <summary>
    /// Delay in seconds before kubelet performing the first probe.
    /// </summary>
    public int InitialDelaySeconds { get; set; }
}

// Identify a "gRPC" health check annotation
internal class GrpcHealthCheckAnnotation(string key, string endpointName, EndpointHealthCheckKind kind)
    : EndpointHealthCheckAnnotation(key, endpointName, kind)
{
}

// Identify a "http" health check annotation
internal class HttpHealthCheckAnnotation(string key, string endpointName, EndpointHealthCheckKind kind, string path)
    : EndpointHealthCheckAnnotation(key, endpointName, kind)
{
    /// <summary>
    /// Path to use for health check.
    /// </summary>
    public string Path { get; } = path;
}
```
In this way we can store information that are missing in the current Health Check Annotation.

Example of how to use during k8s publish
```cs
xxx.PublishAsKubernetesService(resource =>
    {
        // Check for endpoint health check annotation
        if (resource.TryGetAnnotationsOfType<EndpointHealthCheckAnnotation>(out var endpointHealthCheckAnnotations))
        {
            // Create probes
            ProbeV1? readinessProbe = null;
            ProbeV1? livenessProbe = null;
            foreach (var endpointHealthCheckAnnotation in endpointHealthCheckAnnotations)
            {
                // Try get endpoint
                if (!resource.TryGetEndpoints(out var endpoints))
                {
                    // Throw?
                    return;
                }

                var endpoint = endpoints.FirstOrDefault(endpoint => endpoint.Name == endpointHealthCheckAnnotation.EndpointName);
                if (endpoint is null || !endpoint.TargetPort.HasValue)
                {
                    // Throw?
                    return;
                }

                // Prepare probe
                if (endpointHealthCheckAnnotation.Kind == EndpointHealthCheckKind.Readiness)
                {
                    SetProbeDataFromAnnotation(readinessProbe = new());
                }
                else
                {
                    SetProbeDataFromAnnotation(livenessProbe = new());
                }

                void SetProbeDataFromAnnotation(ProbeV1 probe)
                {
                    if (endpointHealthCheckAnnotation is GrpcHealthCheckAnnotation grpcHealthCheckAnnotation)
                    {
                        probe.Grpc = new()
                        {
                            // Use "kind" as "service" because allowed values are the same
                            Service = grpcHealthCheckAnnotation.Kind.ToString().ToLowerInvariant(),
                            Port = endpoint.TargetPort.Value,
                        };
                    }
                    else if (endpointHealthCheckAnnotation is HttpHealthCheckAnnotation httpHealthCheckAnnotation)
                    {
                        probe.HttpGet = new()
                        {
                            Path = httpHealthCheckAnnotation.Path,
                            Scheme = endpoint.UriScheme,
                            Port = endpoint.TargetPort.Value,
                        };
                    }

                    if (endpointHealthCheckAnnotation.InitialDelaySeconds > 0)
                    {
                        probe.InitialDelaySeconds = endpointHealthCheckAnnotation.InitialDelaySeconds;
                    }
                }
            }

            resource.Workload = new Deployment()
            {
                Spec = new()
                {
                    Template = new()
                    {
                        Spec = new()
                        {
                            Containers =
                            {
                                new()
                                {
                                    LivenessProbe = livenessProbe,
                                    ReadinessProbe = readinessProbe,
                                },
                            },
                        },
                    },
                },
            };
        }
    });
```
Some changes have to be done in current `WithHttpHealthCheck` and a new `WithGrpcHealthCheck` will be added (works in the same way but use a grpc health check from `Grpc.HealthCheck` nuget package).
If it would be reasonable I can contribute with PR (so I can remove custom impl in our codebase yeah)

**Comment by davidfowl** [comment]
I don't think this is complete enough for a PR yet. These are health probes not health checks. I would say that it should be possible to define probes (liveness readiness etc) and then they can interact with various environments. 

In [this PR](https://github.com/GMouaad/aspire/pull/1) there's a probe annotaton, why not start from there?

Then we can decide how to turn probes into health checks at runtime *and* how they show up in deployment environments (I see you already have k8s), we need to explore docker compose, and the azure compute environments.

**Comment by Dona278** [comment]
I'll check the PR, adapt with aspire changes occurred in the meanwhile (if required) and write a proposal to integrate with all publisher

**Comment by Dona278** [comment]
I'm back with some points after a little play..
New annotations so we can add specific properties on specific usage:
```cs
/// <summary>
/// Represents an annotation that specifies the probes (health, readiness, liveness, etc.) of a resource
/// </summary>
public abstract class ProbeAnnotation : IResourceAnnotation
{
    /// <summary>
    /// The type of this health probe (startup, readiness or liveness).
    /// </summary>
    public required ProbeType Type { get; init; }

    /// <summary>
    /// The endpoint reference used for the probe
    /// </summary>
    public required EndpointReference EndpointReference { get; init; }

    /// <summary>
    /// The initial delay before the probe should be called
    /// </summary>
    public int InitialDelaySeconds { get; } = 5;

    /// <summary>
    /// The period between each probe call
    /// </summary>
    public int PeriodSeconds { get; } = 5;
}

/// <summary>
/// Represents an annotation that specifies the HTTP probes (health, readiness, liveness, etc.) of a resource
/// </summary>
public sealed class HttpProbeAnnotation : ProbeAnnotation
{
    /// <summary>
    /// The path to the health probe endpoint.
    /// </summary>
    public required string Path { get; init; }
}

/// <summary>
/// Represents an annotation that specifies the gRPC probes (health, readiness, liveness, etc.) of a resource
/// </summary>
public sealed class GrpcProbeAnnotation : ProbeAnnotation
{
    /// <summary>
    /// The service of the health probe endpoint.
    /// </summary>
    public required string Service { get; init; }
}

.... maybe also
TpcProbeAnnotation : ProbeAnnotation
CommandProbeAnnotation : ProbeAnnotation
```

New interface required for extensions and to identity what resource can have probes
```cs
public interface IResourceWithProbes : IResource
```

At this point we need a `.WithProbe()` extension to add annotation, but we have different kind which required different parameters so we can add:
```cs
public static IResourceBuilder<T> WithHttpProbe<T>(this IResourceBuilder<T> builder, ProbeType type, string endpointName, string path, int initialDelaySeconds = 5, int periodSeconds = 5) where T : IResourceWithProbes

public static IResourceBuilder<T> WithGrpcProbe<T>(this IResourceBuilder<T> builder, ProbeType type, string endpointName, string service, int initialDelaySeconds = 5, int periodSeconds = 5) where T : IResourceWithProbes

.... and tcp / command ones
```

For "http" and "grpc" maybe we can implicit add a health check? So the user doesn't had to declare both (with the same parameters) and health checks still be separated from probes, but for these specific ones is most probably that the service was configured to have health check and the user add probes just because want health probe, or not?
Note: A grpc health check need to be added (maybe before this part) to aspire.

For publishers:
- k8s can add probes within `ToContainerV1()` which already use endpoints for ports, so can retrieve probes annotations and if required using endpoints for port and name;
- aca (reading from code because i never used it) can add probes within `BuildContainerApp()` when creates container in the same way of above;
- docker-compose can add probes within `CreateDockerComposeServiceResourceAsync()`  in the same way as above BUT support only one "health check" [(docs)](https://docs.docker.com/reference/compose-file/services/#healthcheck)

**Comment by davidfowl** [comment]
The base ProbeAnnotation shouldn't have an EndpointReference on it. This looks like a good MVP. Let's just start with HttpProbeAnnotation for now and get that working end to end before we expand to gRPC and others.

**Comment by Dona278** [comment]
Ok! Maybe next a "EndpointProbeAnnotation" as a base intermediate layer? You know I've always wanted grpc lol

**Comment by davidfowl** [comment]
Grpc will be trivial to do once we understand the pattern 😄.  I'd love to avoid a big hierarchy of types that serve little purpose but it seems like EndpointProbeAnnotation might be useful base type for Http/Tcp/Grpc. Maybe you only need ProbeAnnotation and EndpointProbeAnnotation for now.

**Comment by Dona278** [comment]
Manifest (aspire) can be ignored since will be deprecated? Otherwise "probes" can be added under the resource?

**Comment by davidfowl** [comment]
Yes let’s ignore the manifest

**Comment by Dona278** [comment]
Created a [PR](https://github.com/dotnet/aspire/pull/11081)
It is draft because I have some doubts:
- I've implemented only http probes for now, let me know it the api is ok
- docker compose allow only one probe called "health check", how we have to manage this behavior? Currently will be used the first one by prioritize "liveness", I don't know if we want to throw exception if there are more than one probe associated to the resource;
- I allowed probe on container and project, wait for confirm
- I hardcoded "10s" as timeout for docker-compose but I'll add "Timeout" as property tomorrow (I forgot that)

**Comment by davidfowl** [comment]
I think we should skip the docker compose implementation for now as it’s not bulletproof. Curl may not even be available in the container so we may need this to be more opt in than it is today.

**Comment by Dona278** [comment]
Totally agree, however user may use probes annotation and customize publish himself, when "CommandProbe" will be available I think we could check for more than 1 or for probe different from this one and throw notsuppurted beacuse docker compose healthcheck seems to work only with command.
